design/annotation-for-power-cycling-and-deleting-failed-nodes.md

```diff
@@ -0,0 +1,180 @@
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+It is not always practical to require admin intervention once a node has been
+identified as having reached a bad or unknown state.
+
+In order to automate the recovery of exclusive workloads (eg. RWO volumes and
+StatefulSets), we need a way to put failed nodes into a safe state, indicate to
+the scheduler that affected workloads can be started elsewhere, and then
+attempt to recover capacity.
+
+## Motivation
+
+Hardware is imperfect, and software contains bugs. When node level failures
+such as kernel hangs or dead NICs occur, the work required from the cluster
+does not decrease - workloads from affected nodes need to be restarted
+somewhere. 
+
+However some workloads may require at-most-one semantics.  Failures affecting
+these kind of workloads risk data loss and/or corruption if "lost" nodes are
+assumed to be dead when they are still running.  For this reason it is
+important to know that the node has reached a safe state before initiating
+recovery of the workload.
+
+Powering off the affected node via IPMI or related technologies achieves this,
+but must be paired with deletion of the Node object to signal to the scheduler
+that no Pods or PersistentVolumes are present there.
+
+Ideally customers would over-provision the cluster so that a node failure (or
+several) does not put additional stress on surviving peers, however budget
+constraints mean that this is often not the case, particularly in Edge
+deployments which may consist of as few as three nodes of commodity hardware. 
+Even when deployments start off over-provisioned, there is a tendency for the
+extra capacity to become permanently utilised.  It is therefore usually
+important to recover the lost capacity quickly.
+
+For this reason it is desirable to power the machine back on again, in the hope
+that the problem was transient and the node can return to a healthy state.
+Upon restarting, kubelet automatically contacts the masters to re-register
+itself, allowing it to host workloads.
+
+### Goals
+
+- Enable the safe automated recovery of exclusive workloads
+- Provide the possibility of restoring cluster capacity automatically
+
+### Non-Goals
+
+- Creating a public API for rebooting a node
+- Identifying when automated recovery should be triggered
+- Taking actions other than power cycling to return the node to full health
+
+## Proposal
+
+This proposal calls for a new controller which watches for the presence of a
+specific Node annotation.  If present, the controller will locate the Machine
+and BareMetalHost host objects via their annotations, and (in serial) use the
+BareMetalHost API to power off the machine, delete the Node, and then power the
+machine back on again.
+
+In order to track the state of the recovery workflow, a [new
+CRD](https://github.com/kubevirt/machine-remediation/blob/master/pkg/apis/machineremediation/v1alpha1/machineremediation_types.go)
+is proposed.
```

Comment:
Is it the same controller that both watches for the annotation and reconciles the new CRD?
I ask because if we implement the proposal in #48 then we likely won't need this CRD any longer, but we would still need the controller to watch the Node, so it would be good to make them easily separable.

Reply:
There are two separate controllers: nodereboot and machineremediation
So I think we're prepared for that

Reply:
Updated proposal with details